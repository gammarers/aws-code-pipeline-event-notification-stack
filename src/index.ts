import * as crypto from 'crypto';
import * as cdk from 'aws-cdk-lib';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import * as sns from 'aws-cdk-lib/aws-sns';
import * as subscriptions from 'aws-cdk-lib/aws-sns-subscriptions';
import * as sfn from 'aws-cdk-lib/aws-stepfunctions';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Construct } from 'constructs';

export interface NotificationsProperty {
  readonly emails?: string[];
}
export interface CodePipelineEventNotificationStackProps extends cdk.StackProps {
  readonly notifications: NotificationsProperty;
}

export class CodePipelineEventNotificationStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CodePipelineEventNotificationStackProps) {
    super(scope, id, props);

    const random: string = crypto.createHash('shake256', { outputLength: 4 })
      .update(cdk.Names.uniqueId(this))
      .digest('hex');

    // SNS Topic for notifications
    const topic: sns.Topic = new sns.Topic(this, 'CodePipelineNotificationTopic', {
      topicName: `code-pipeline-event-notification-${random}-topic`,
      displayName: 'CodePipeline Event Notification Topic',
    });

    // Subscribe an email endpoint to the topic
    for (const email of props.notifications.emails ?? []) {
      topic.addSubscription(new subscriptions.EmailSubscription(email));
    }

    // Subscribe a HTTP endpoint (Slack Webhook) to the topic
    //topic.addSubscription(new subs.UrlSubscription('https://hooks.slack.com/services/YOUR/SLACK/WEBHOOK'));

    const succeed: sfn.Succeed = new sfn.Succeed(this, 'Succeed');

    // Step Functions Tasks
    const getResourceTagMappingList: tasks.CallAwsService = new tasks.CallAwsService(this, 'GetResourceTagMappingList', {
      service: 'resourcegroupstaggingapi',
      action: 'getResources',
      parameters: {
        ResourceARNList: sfn.JsonPath.listAt('$.resources'),
        //        TagFilters: [
        //          {
        //            Key: 'DeployNotification',
        //            Values: ['YES'],
        //          },
        //        ],
        // ResourceTypeFilters: ['codepipeline:pipeline'], // ResourceTypeFilters is not allowed when providing a ResourceARNList
      },
      iamAction: 'tag:GetResources',
      iamResources: ['*'],
      //resultPath: '$.tagsResult',
      resultPath: '$.Result.GetResource',
      //      resultSelector: {
      //        'Tags.$': '$.ResourceTagMappingList[0].Tags',
      //      },
    });
    //getTags.addCatch()

    const findTagVluesPass: sfn.Pass = new sfn.Pass(this, 'FindTagVlues', {
      parameters: {
        //'Find.$': "States.JsonToString($.tagsResult.Tags[?(@.Key == 'DeployNotification')])",
        //Found: "States.JsonToString($.tagsResult.Tags[?(@.Key == 'DeployNotification')])",
        Found: sfn.JsonPath.stringAt("$.Result.GetTags.Tags[?(@.Key == 'DeployNotification')].Value"),
        //Find: sfn.JsonPath.stringToJson('States.JsonToString($.tagsResult.Tags[?(@.Key == "DeployNotification")])'),
        //Find: sfn.JsonPath.jsonToString("$.tagsResult.Tags[?(@.Key == 'DeployNotification')]"),
      },
      resultPath: '$.Result.FindTagValues',
    });

    const checkResouceTagsExist: sfn.Choice = new sfn.Choice(this, 'CheckResouceTagsExist')
      .when(sfn.Condition.isPresent('$.Result.GetResource.ResourceTagMappingList[0].Tags'),
        new sfn.Pass(this, 'TagsExist', {
          parameters: {
            Tags: sfn.JsonPath.stringAt('$.Result.GetResource.ResourceTagMappingList[0].Tags'),
          },
          resultPath: '$.Result.GetTags',
        }).next(findTagVluesPass))
      .otherwise(new sfn.Pass(this, 'NoTagsFound', {
        comment: 'This Reouces is not set Tags',
      }));

    getResourceTagMappingList.next(checkResouceTagsExist);
    //    const checkIfFoundState = new sfn.Choice(this, 'CheckFindTag')
    //      .when(sfn.Condition.isPresent('$.findTag.Found'), new sfn.Succeed(this, 'BBBSucceed'))
    //      .otherwise(new sfn.Pass(this, 'TagNotFound'));

    //const checkArrayContains = sfn.Condition.booleanEquals('$.arrayContainsResult', true);

    // 配列に値が含まれているかどうかを計算するPassステート
    const checkArrayContainsPass: sfn.Pass = new sfn.Pass(this, 'CheckArrayContains', {
      parameters: {
        Is: sfn.JsonPath.arrayContains(sfn.JsonPath.stringAt('$.Result.FindTagValues.Found'), 'YES'),
      },
      resultPath: '$.Result.ContainTag',
    });

    findTagVluesPass.next(checkArrayContainsPass);

    const messageStatusIcons = {
      started: '🥳',
      succeeded: '🤩',
      resumed: '🤔',
      failed: '😫',
      stopping: '😮',
      stopped: '😌',
      superseded: '🧐',
    };

    const createPreparePipelineMessageSubject = ((statusIcon: string) => {
      return sfn.JsonPath.format(`${statusIcon} [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]`,
        sfn.JsonPath.stringAt('$.detail.state'),
        sfn.JsonPath.stringAt('$.account'),
        sfn.JsonPath.stringAt('$.region'),
      );
    });

    const preparePipelineMessage: sfn.Pass = new sfn.Pass(this, 'PrepareStartedPipelineMessage', {
      parameters: {
        Subject: createPreparePipelineMessageSubject(messageStatusIcons.started),
        Message: sfn.JsonPath.format('Account : {}\nRegion : {}\nPipeline :  {}\nState : {}',
          sfn.JsonPath.stringAt('$.account'),
          sfn.JsonPath.stringAt('$.region'),
          sfn.JsonPath.stringAt('$.detail.pipeline'),
          sfn.JsonPath.stringAt('$.detail.state'),
        ),
      },
    });

    const checkPipelineStateMatch: sfn.Choice = new sfn.Choice(this, 'CheckPipelineStateMatch')
      .when(sfn.Condition.stringEquals('$.detail.state', 'STARTED'), preparePipelineMessage)
      .when(sfn.Condition.stringEquals('$.detail.state', 'SUCCEEDED'), preparePipelineMessage)
      .when(sfn.Condition.stringEquals('$.detail.state', 'RESUMED'), preparePipelineMessage)
      .when(sfn.Condition.stringEquals('$.detail.state', 'FAILED'), preparePipelineMessage)
      .when(sfn.Condition.stringEquals('$.detail.state', 'STOPPING'), preparePipelineMessage)
      .when(sfn.Condition.stringEquals('$.detail.state', 'STOPPED'), preparePipelineMessage)
      .when(sfn.Condition.stringEquals('$.detail.state', 'SUPERSEDED'), preparePipelineMessage)
      .otherwise(new sfn.Pass(this, 'NoMatchValue'));

    const checkFoundTagMatch = new sfn.Choice(this, 'CheckFoundTagMatch')
      .when(sfn.Condition.booleanEquals('$.Result.ContainTag.Is', true), checkPipelineStateMatch)
      .otherwise(new sfn.Pass(this, 'NoMatchPipelineState'));

    checkArrayContainsPass.next(checkFoundTagMatch);

    const sendNotification: tasks.SnsPublish = new tasks.SnsPublish(this, 'SendNotification', {
      topic: topic,
      subject: sfn.JsonPath.stringAt('$.Subject'),
      message: sfn.TaskInput.fromJsonPathAt('$.Message'),
      resultPath: '$.snsResult',
    });

    preparePipelineMessage.next(sendNotification);

    sendNotification.next(succeed);

    // Step Functions State Machine
    const stateMachine: sfn.StateMachine = new sfn.StateMachine(this, 'StateMachine', {
      stateMachineName: `codepipeline-event-notification-${random}-state-machine`,
      timeout: cdk.Duration.minutes(5),
      definition: getResourceTagMappingList,
    });

    // EventBridge Rule
    new events.Rule(this, 'Rule', {
      ruleName: `codepipeline-event-catch-${random}-rule`,
      eventPattern: {
        source: ['aws.codepipeline'],
        detailType: ['CodePipeline Pipeline Execution State Change'],
      },
      targets: [new targets.SfnStateMachine(stateMachine)],
    });
  }
}
