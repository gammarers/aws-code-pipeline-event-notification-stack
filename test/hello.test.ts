import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CodePipelineEventNotificationStack } from '../src';

test('hello', () => {
  const app = new App();

  const stack = new CodePipelineEventNotificationStack(app, 'CodePipelineEventNotificationStack', {
    notifications: {
      emails: [
        'foo@example.com',
      ],
    },
  });

  const template = Template.fromStack(stack);

  expect(template.toJSON()).toMatchSnapshot();
});