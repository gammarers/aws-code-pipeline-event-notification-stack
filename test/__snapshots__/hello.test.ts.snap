// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hello 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CodePipelineNotificationTopic51FFA519": {
      "Properties": {
        "DisplayName": "CodePipeline Event Notification Topic",
        "TopicName": "code-pipeline-event-notification-ecebfe7c-topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "CodePipelineNotificationTopicfooexamplecomA65A6538": {
      "Properties": {
        "Endpoint": "foo@example.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "CodePipelineNotificationTopic51FFA519",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "Rule4C995B7F": {
      "Properties": {
        "EventPattern": {
          "detail-type": [
            "CodePipeline Pipeline Execution State Change",
          ],
          "source": [
            "aws.codepipeline",
          ],
        },
        "Name": "codepipeline-event-catch-ecebfe7c-rule",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "StateMachine2E01A3A5",
            },
            "Id": "Target0",
            "RoleArn": {
              "Fn::GetAtt": [
                "StateMachineEventsRoleDBCDECD1",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "StateMachine2E01A3A5": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"GetResourceTagMappingList","States":{"GetResourceTagMappingList":{"Next":"CheckResouceTagsExist","Type":"Task","ResultPath":"$.Result.GetResource","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::aws-sdk:resourcegroupstaggingapi:getResources","Parameters":{"ResourceARNList.$":"$.resources"}},"CheckResouceTagsExist":{"Type":"Choice","Choices":[{"Variable":"$.Result.GetResource.ResourceTagMappingList[0].Tags","IsPresent":true,"Next":"TagsExist"}],"Default":"NoTagsFound"},"NoTagsFound":{"Type":"Pass","Comment":"This Reouces is not set Tags","End":true},"TagsExist":{"Type":"Pass","ResultPath":"$.Result.GetTags","Parameters":{"Tags.$":"$.Result.GetResource.ResourceTagMappingList[0].Tags"},"Next":"FindTagVlues"},"FindTagVlues":{"Type":"Pass","ResultPath":"$.Result.FindTagValues","Parameters":{"Found.$":"$.Result.GetTags.Tags[?(@.Key == 'DeployNotification')].Value"},"Next":"CheckArrayContains"},"CheckArrayContains":{"Type":"Pass","ResultPath":"$.Result.ContainTag","Parameters":{"Is.$":"States.ArrayContains($.Result.FindTagValues.Found, 'YES')"},"Next":"CheckFoundTagMatch"},"CheckFoundTagMatch":{"Type":"Choice","Choices":[{"Variable":"$.Result.ContainTag.Is","BooleanEquals":true,"Next":"CheckPipelineStateMatch"}],"Default":"NoMatchPipelineState"},"NoMatchPipelineState":{"Type":"Pass","End":true},"CheckPipelineStateMatch":{"Type":"Choice","Choices":[{"Variable":"$.detail.state","StringEquals":"STARTED","Next":"PrepareStartedPipelineMessage"},{"Variable":"$.detail.state","StringEquals":"SUCCEEDED","Next":"PrepareSucceededdPipelineMessage"},{"Variable":"$.detail.state","StringEquals":"RESUMED","Next":"PrepareResumedPipelineMessage"},{"Variable":"$.detail.state","StringEquals":"FAILED","Next":"PrepareFailedPipelineMessage"},{"Variable":"$.detail.state","StringEquals":"STOPPING","Next":"PrepareStoppingPipelineMessage"},{"Variable":"$.detail.state","StringEquals":"STOPPED","Next":"PrepareStoppedPipelineMessage"},{"Variable":"$.detail.state","StringEquals":"SUPERSEDED","Next":"PrepareSupersededPipelineMessage"}],"Default":"NoMatchValue"},"NoMatchValue":{"Type":"Pass","End":true},"PrepareStartedPipelineMessage":{"Type":"Pass","Parameters":{"Subject.$":"States.Format('ü•≥ [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', $.detail.state, $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline :  {}\\nState : {}', $.account, $.region, $.detail.pipeline, $.detail.state)"},"Next":"SendNotification"},"SendNotification":{"Next":"Succeed","Type":"Task","ResultPath":"$.snsResult","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish","Parameters":{"TopicArn":"",
              {
                "Ref": "CodePipelineNotificationTopic51FFA519",
              },
              "","Message.$":"$.Message","Subject.$":"$.Subject"}},"PrepareSucceededdPipelineMessage":{"Type":"Pass","Parameters":{"Subject.$":"States.Format('ü§© [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', $.detail.state, $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline :  {}\\nState : {}', $.account, $.region, $.detail.pipeline, $.detail.state)"},"Next":"SendNotification"},"PrepareResumedPipelineMessage":{"Type":"Pass","Parameters":{"Subject.$":"States.Format('ü§î [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', $.detail.state, $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline :  {}\\nState : {}', $.account, $.region, $.detail.pipeline, $.detail.state)"},"Next":"SendNotification"},"PrepareFailedPipelineMessage":{"Type":"Pass","Parameters":{"Subject.$":"States.Format('üò´ [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', $.detail.state, $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline :  {}\\nState : {}', $.account, $.region, $.detail.pipeline, $.detail.state)"},"Next":"SendNotification"},"PrepareStoppingPipelineMessage":{"Type":"Pass","Parameters":{"Subject.$":"States.Format('üòÆ [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', $.detail.state, $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline :  {}\\nState : {}', $.account, $.region, $.detail.pipeline, $.detail.state)"},"Next":"SendNotification"},"PrepareStoppedPipelineMessage":{"Type":"Pass","Parameters":{"Subject.$":"States.Format('üòå [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', $.detail.state, $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline :  {}\\nState : {}', $.account, $.region, $.detail.pipeline, $.detail.state)"},"Next":"SendNotification"},"PrepareSupersededPipelineMessage":{"Type":"Pass","Parameters":{"Subject.$":"States.Format('üßê [{}] AWS CodePipeline Pipeline Execution State Notification [{}][{}]', $.detail.state, $.account, $.region)","Message.$":"States.Format('Account : {}\\nRegion : {}\\nPipeline :  {}\\nState : {}', $.account, $.region, $.detail.pipeline, $.detail.state)"},"Next":"SendNotification"},"Succeed":{"Type":"Succeed"}},"TimeoutSeconds":300}",
            ],
          ],
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
        "StateMachineName": "codepipeline-event-notification-ecebfe7c-state-machine",
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "StateMachineEventsRoleDBCDECD1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineEventsRoleDefaultPolicyFB602CA9": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "StateMachine2E01A3A5",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineEventsRoleDefaultPolicyFB602CA9",
        "Roles": [
          {
            "Ref": "StateMachineEventsRoleDBCDECD1",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "StateMachineRoleB840431D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "tag:GetResources",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "CodePipelineNotificationTopic51FFA519",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": [
          {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
